File: apps/demo-be-jasb/src/main/java/com/organiclever/be/auth/repository/UserRepository.java
package com.organiclever.be.auth.repository;

import com.organiclever.be.auth.model.User;
import java.util.Optional;
import java.util.UUID;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface UserRepository extends JpaRepository<User, UUID> {
    Optional<User> findByUsername(String username);

    boolean existsByUsername(String username);

    Optional<User> findByEmail(String email);

    Page<User> findAll(Pageable pageable);

    @Query("SELECT u FROM User u WHERE u.email = :email ORDER BY u.createdAt ASC")
    Page<User> findAllByEmailContaining(@Param("email") String email, Pageable pageable);
}

File: apps/a-demo-be-kotlin-ktor/src/main/kotlin/com/ademobektkt/infrastructure/repositories/UserRepository.kt
package com.ademobektkt.infrastructure.repositories

import com.ademobektkt.domain.Page
import com.ademobektkt.domain.User
import com.ademobektkt.domain.UserStatus
import java.util.UUID

data class CreateUserRequest(
  val username: String,
  val email: String,
  val displayName: String,
  val passwordHash: String,
  val role: com.ademobektkt.domain.Role = com.ademobektkt.domain.Role.USER,
)

data class UpdateUserPatch(
  val displayName: String? = null,
  val passwordHash: String? = null,
  val status: UserStatus? = null,
  val failedLoginAttempts: Int? = null,
  val role: com.ademobektkt.domain.Role? = null,
)

interface UserRepository {
  suspend fun findByUsername(username: String): User?

  suspend fun findById(id: UUID): User?

  suspend fun findByEmail(email: String): User?

  suspend fun create(request: CreateUserRequest): User

  suspend fun update(id: UUID, patch: UpdateUserPatch): User?

  suspend fun incrementFailedLogins(id: UUID): Int

  suspend fun resetFailedLogins(id: UUID)

  suspend fun findAll(page: Int, pageSize: Int, searchFilter: String?): Page<User>
}
